### PR TITLE
Refine leastsq

### PR DESCRIPTION
**Pull request by caspervdw**
This PR builds onto #406 . I implemented all the code for:
- combined linking & feature finding
- separating images in clusters of features
- least squares refinement of overlapping features, using any radial model function
- adapted `artificial.py` and coupled tests in `test_feature` and `test_find` to work with a better convention for artificial feature creation

The description of the algorithm is available online in https://arxiv.org/abs/1607.08819. A revised version is currently ``provisionally accepted'' by a peer-reviewed journal. Will keep you updated about how that is going.

I am referring to trackpy in the paper in the sense that it ''will be implemented in trackpy, but currently available on https://github.com/caspervdw/clustertracking''. It would be neater to change that to ''will be  available from trackpy v0.4'', as I think it might pull some more users to trackpy. Is this OK with you guys? I will probably have to supply the definitive version of the paper in the next week or so. It's not that this needs to be merged by then, just the intention of putting it inside v0.4 is enough.

### CODE REVIEW

**Comment by caspervdw** [comment]
I finished the code and docstrings. Let's get in touch about how to proceed. @danielballan @nkeim @tacaswell 

Merging the refine refactor in #406 would certainly make this PR more manageable, but still, this PR contains 5000 new lines which will take a lot of time to review.

**Comment by danielballan** [comment]
This seems like as good an excuse as any to set up a core developer group video chat. Would people be up for that next week? @nkeim @tacaswell?

**Comment by caspervdw** [comment]
Thanks for merging #406, I rebased this PR on master.
A group video chat would be great! Let's pick a time over email.

**Comment by nkeim** [comment]
Yes, sorry I've been absent for so long! The worst is of that ended last week.

I think Casper and I are 9 time zones apart. So if I talked at 8 am here, that would be 11 for Dan and Tom, and 5 pm for Casper. Next week I will be available at 8am every day but Thursday. Tuesday would be especially good.

Best,
Nathan

On Nov 10, 2016, at 9:34 AM, Casper van der Wel <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks for merging #406<https://github.com/soft-matter/trackpy/pull/406>, I rebased this PR on master.
A group video chat would be great! Let's pick a time over email.

**Comment by caspervdw** [comment]
Up to 9pm is also ok with me, which is noon for nathan and 3pm for Dan and Tom. Tuesday would work for me too.

**Comment by caspervdw** [comment]
Is tomorrow at 4pm UTC ok with everyone?

**Comment by nkeim** [comment]
Still works for me.

On Nov 14, 2016, at 10:31 AM, Casper van der Wel <notifications@github.com<mailto:notifications@github.com>> wrote:


Is tomorrow at 4pm UTC ok with everyone?

**Comment by danielballan** [comment]
Hey, sorry, I have a meeting with some higher-ups at that time that I can't easily reschedule. It won't take long; I can be free between 5pm and 7pm UTC (12-2pm EST).

**Comment by caspervdw** [comment]
@nkeim About the merge of the linking 'internal forks': I can write a function `link_simple_iter` and `link_simple_df` that does linking with the reimplemented code in this PR. That could be a starting point for the port of the adaptive linking and prediction. It might be worthwhile to keep the `simple` versions because it might be a bit faster.

**Comment by caspervdw** [comment]
@nkeim @danielballan @tacaswell  all done, this can be merged. I did touch some code in `linking.py` in the process of rebasing my `find_link` codebase. In particular, `search_range` can now be anisotropic (sticking to my own tradition). The coordinates get rescaled with the `search_range` at a certain point, so that `search_range=1` for the deep-down linking code (e.g. `SubnetLinker`).

I made thin wrappers of my new `find_link.Linker` class, one that takes an iterable of numpy arrays and returns a generator of particle-id-lists (`link_simple_iter`), and one that takes a DataFrame and returns a DataFrame with a `'particle'` column added (`link_simple`). All the routines that are based on iterable of `Point` classes (`link`, `link_iter`) are too obscure for most users to understand (I am actually citing a colleague now). 

@nkeim A next step would be copying over everything to a subpackage `trackpy.linking`, and rebasing the contents of `linking.py` on a common `Linker` class. I think that adaptive linking, prediction, and diagnostics should have their own `Linker` classes, for code understandability and also for performance.

Next to that, I rewrote the `introduction.rst` a bit to reflect the recent refactor into sub-packages.

Before merging, I hope somebody can have a look at `linking.py` and `introduction.rst`.

**Comment by danielballan** [comment]
Great! I can't personally look at this today but will try hard to get to it early this week.

**Comment by caspervdw** [comment]
Just published! :smiley:
Can't put it on arXiv for a year due to IOPs copyright regulations. I will update the docstring with the proper reference.
http://iopscience.iop.org/article/10.1088/1361-648X/29/4/044001

**Comment by danielballan** [comment]
Congrats on the publication! I set a reminder to look at linking.py this weekend.

File: trackpy/linking.py [hunk 1]
```diff
@@ -1072,6 +1141,12 @@ def link(self, levels):
                 if sp is not None:
                     del sp.forward_cands
 
+            # TODO: Emit debug message with number of
+            # subnets in this level, numbers of new/remembered/lost particles
```
Comment (danielballan) [inline trackpy/linking.py]: Good idea.

File: doc/introduction.rst [hunk 2]
```diff
@@ -21,26 +22,30 @@ Features
...
+we separate *tracking* into three separate steps. In the first step, *feature finding*
+initial feature coordinates are obtained from the images. Subsequently, sub-pixel precision
+is obtained in coordinate *refinement*. Finally, the coordaintes are *linked* in time yielding
+the feature trajectories.
+
+  * The tracking algorithm originally implemented by John Crocker and Eric Weeks in IDL was
+    completely reimplemented in Python.
+  * A `flexible framework for least-squares fitting <https://arxiv.org/abs/1607.08819>`__
+    allows for sub-pixel refinement using any radial model function in 2D and 3D.
+  * Trackpy actively used and tested on **Windows, Mac OSX, and Linux**,
```
Comment (danielballan) [inline doc/introduction.rst]: 'Trackpy actively' -> 'Trackpy is actively'

File: doc/introduction.rst [hunk 3]
```diff
@@ -106,7 +113,8 @@ Core Contributors
   * **Nathan Keim** alternative trajectory-linking implementations, major
     speed-ups, prediction, adaptive search
   * **Thomas Caswell** multiple implementations of sophisticated trajectory-linking, tests
-  * **Casper van der Wel** anisotropic (3D) feature-finding, 3D plotting and analyses
+  * **Casper van der Wel** anisotropic 3D feature-finding, plotting and analyses, framework
```
Comment (danielballan) [inline doc/introduction.rst]: Feel free to put your name first in the list as the current most active maintainer.

File: trackpy/linking.py (hunk 1 above)
Comment (caspervdw) [inline trackpy/linking.py]: Not my idea, I just moved the comment together with the `yield`.

**Comment by caspervdw** [comment]
Thanks @danielballan for the review and merge! Let's plan a minor release soon.

**Pull request merged**